[scroll-animations] WPT tests `scroll-timelines/setting-start-time.html` and `scroll-timelines/setting-current-time.html` under `scroll-animations` are crashes

graouts opened the pull request:
#### be9c3ff059a105eba308b1127545802fea2f8280
<pre>
[scroll-animations] WPT tests `scroll-timelines/setting-start-time.html` and `scroll-timelines/setting-current-time.html` under `scroll-animations` are crashes
<a href="https://bugs.webkit.org/show_bug.cgi?id=281731">https://bugs.webkit.org/show_bug.cgi?id=281731</a>
<a href="https://rdar.apple.com/138170672">rdar://138170672</a>

Reviewed by Tim Nguyen.

We would incorrectly mark absolute time values provided through the bindings
for `Animation.currentTime` and `Animation.startTime` as valid because we failed
to validate them per spec accounting for the associated timeline type (progress-based
vs. monotonic). This meant we&apos;d let some absolute time values slip by and later
compare them to percentage time values which would trigger some debug assertions
and a crash.

We now correctly report any `CSSNumberishTime` that was successfully created from
a `CSSNumberish` as valid through `CSSNumberishTime::isValid()` and check whether
the provided type matches the current timeline in `WebAnimation::isTimeValid()`.

We also update the tests in question to expect a JS `TypeError` rather than a DOM
`NotSupported` since that&apos;s what the Web Animations spec discusses in
<a href="https://drafts.csswg.org/web-animations-2/#validating-a-css-numberish-time.">https://drafts.csswg.org/web-animations-2/#validating-a-css-numberish-time.</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-current-time-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-current-time.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-start-time-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-start-time.html:
* Source/WebCore/animation/CSSNumberishTime.cpp:
(WebCore::CSSNumberishTime::isValid const):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::isTimeValid const):
(WebCore::WebAnimation::setBindingsStartTime):
(WebCore::WebAnimation::setBindingsCurrentTime):
* Source/WebCore/animation/WebAnimation.h:

Canonical link: <a href="https://commits.webkit.org/285390@main">https://commits.webkit.org/285390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc20ccb385150d6a27c010ef40fad5cfad532356

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51951 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25318 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76714 "Built successfully") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23739 "Failed to checkout and rebase branch from PR 35417") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74645 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23561 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/23739 "Failed to checkout and rebase branch from PR 35417") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75597 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/47023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/62458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/43674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/19927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22089 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/20283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16771 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/78383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16819 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/62471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/13100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11129 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47749 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48818 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50108 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48561 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -363,9 +363,21 @@ void WebAnimation::effectTargetDidChange(const std::optional<const Styleable>& p
+    // https://drafts.csswg.org/web-animations-2/#validating-a-css-numberish-time
+    if (time && !time->isValid())
+        return false;
+    if (m_timeline && m_timeline->isProgressBased() && time && time->time())
+        return false;
+    if ((!m_timeline || m_timeline->isMonotonic()) && time && time->percentage())
```
Why does the spec express the last check as negation? 

e.g. "Is not a duration" instead of "is a percentage" 

Is there a specific reason? or is spec language just unnecessarily complicated?

graouts replied:
Good question, but I don't know.

graouts committed:
[scroll-animations] WPT tests `scroll-timelines/setting-start-time.html` and `scroll-timelines/setting-current-time.html` under `scroll-animations` are crashes
https://bugs.webkit.org/show_bug.cgi?id=281731
rdar://138170672

Reviewed by Tim Nguyen.

We would incorrectly mark absolute time values provided through the bindings
for `Animation.currentTime` and `Animation.startTime` as valid because we failed
to validate them per spec accounting for the associated timeline type (progress-based
vs. monotonic). This meant we'd let some absolute time values slip by and later
compare them to percentage time values which would trigger some debug assertions
and a crash.

We now correctly report any `CSSNumberishTime` that was successfully created from
a `CSSNumberish` as valid through `CSSNumberishTime::isValid()` and check whether
the provided type matches the current timeline in `WebAnimation::isTimeValid()`.

We also update the tests in question to expect a JS `TypeError` rather than a DOM
`NotSupported` since that's what the Web Animations spec discusses in
https://drafts.csswg.org/web-animations-2/#validating-a-css-numberish-time.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-current-time-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-current-time.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-start-time-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/scroll-timelines/setting-start-time.html:
* Source/WebCore/animation/CSSNumberishTime.cpp:
(WebCore::CSSNumberishTime::isValid const):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::isTimeValid const):
(WebCore::WebAnimation::setBindingsStartTime):
(WebCore::WebAnimation::setBindingsCurrentTime):
* Source/WebCore/animation/WebAnimation.h:

Canonical link: https://commits.webkit.org/285390@main
